programs/stake/src/stake_state.rs

```diff
@@ -836,17 +877,18 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
 }
 
 // utility function, used by runtime
+// returns a tuple of (stakers_reward,voters_reward)
 pub fn redeem_rewards(
     stake_account: &mut Account,
     vote_account: &mut Account,
-    point_value: f64,
+    point_value: &PointValue,
     stake_history: Option<&StakeHistory>,
 ) -> Result<(u64, u64), InstructionError> {
     if let StakeState::Stake(meta, mut stake) = stake_account.state()? {
         let vote_state: VoteState =
             StateMut::<VoteStateVersions>::state(vote_account)?.convert_to_current();
 
-        if let Some((voters_reward, stakers_reward)) =
+        if let Some((stakers_reward, voters_reward)) =
             stake.redeem_rewards(point_value, &vote_state, stake_history)
         {
             stake_account.lamports += stakers_reward;
```

Comment:
nits: btw https://github.com/solana-labs/solana/pull/10914/files#diff-0c1e264efc2089624e55f89513c862b8R897 needs to be `.unwrap()` and incremented `.lamports`. Otherwise, we might hide serious issues in the code, resulting in silently fail to deliver legitimate rewards to stakers because we're silently discarding errors for that accounts. 

Before reaching to that code, we've only done `redeem_rewards()` (note: doesn't return `Result`/`Option`; should always succeed.) after successfully restoring the stake state.

Reply:
`redeem_rewards()` returns `None` in the case where no rewards are paid, to indicate that `credits_observed` shouldn't be changed

Reply:
@rwalker-com Sorry, my comment was misleading.. I commented about the `?` in  `stake_account.set_state(...)`_**`?`**_ here:

https://github.com/solana-labs/solana/blob/d19cd3a18aee0a5cb859bba6aac07cd10e1af508/programs/stake/src/stake_state.rs#L855

I'm aware the case `redeem_rewards()` returning `None` is legitimate as you mentioned and I think it's not affected by converting the `?` on `.set_state()` to `unrwap()`.

Reply:
I think changing the '?' to '.unwrap()' might be problematic unless the `redeem_rewards()` instruction is no longer valid.

The error is not silently ignored.  Any failure in this path should prevent any `store_account()`.

Reply:
Thanks for explaining further! Bear with me for following questions because I'm still new to the `stake` native program...

> I think changing the `?` to `.unwrap()` might be problematic unless the `redeem_rewards()` instruction is no longer valid.

Yeah, I think it's problematic if the instruction is still valid. At the same time, I think `RedeemVoteCredits` instruction is no longer valid since #7916. Assuming you're probably worried about `.unwrap()` on untrusted user-supplied transaction, is there something else that's open to attacks still? I thought the only valid call-path to this `redeem_rewards` function is now from `Bank::update_rewards()`, which can be regarded as processing on trusted data set.

Also, https://github.com/solana-labs/solana/blob/d19cd3a18aee0a5cb859bba6aac07cd10e1af508/programs/stake/src/stake_state.rs#L855 is ensured to `set_state()` something that is just successfully loaded from `StakeState::Stake(meta, mut stake)`.



> The error is not silently ignored. Any failure in this path should prevent any `store_account()`.

Yeah,  `store_account()` of both the stake and the vote _for that particular delegation_ isn't stored. For other delegations, they just succeeds because we're continuing the outer loop while ignoring the error: https://github.com/solana-labs/solana/blob/d19cd3a18aee0a5cb859bba6aac07cd10e1af508/runtime/src/bank.rs#L956-L962

I agree it's ok to continue the loop without doing `store_account()` in the case where no rewards are paid for some delegations. But I think it's problematic to ignore something that shouldn't occur ever (= bad state transition in `Stake::redeem_accounts(&mut self)`, where it's only setting `self.credits_ovserved` and incrementing `self.delegation.stake` by `stakers_reward`).